Implement metadata for plugins

Ohisemega opened the pull request:
Implement metadata for BloscPlugin.
Add update to BloscPluginTestFixture to check for Metadata.

Fixes #408

GDYendell commented:
This need rebasing once #427 is merged

Ohisemega committed:
Add ParamMetadata helpers and overrides without min/max

Ohisemega committed:
Add param metadata to blosc plugin

Ohisemega committed:
Merge branch 'master' into Implement_Core_Plugins_Metadata_Fix_ISSUE_408

Ohisemega committed:
Update add_*_metadata() API Calls

Ohisemega committed:
Merge branch 'master' into Implement_Core_Plugins_Metadata_Fix_ISSUE_408

Ohisemega committed:
Merge branch 'master' into Implement_Core_Plugins_Metadata_Fix_ISSUE_408

Ohisemega committed:
Add Simplified Aliases to Base Plugin

Add Simplified Aliases to Base Plugin - FrameProcessorPlugin, to be used by derived member classes.

GDYendell commented on cpp/frameProcessor/src/BloscPlugin.cpp:
```diff
@@ -55,6 +55,16 @@ void create_cd_values(const BloscCompressionSettings& settings, std::vector<unsi
+  add_status_param_metadata(CONFIG_BLOSC_COMPRESSOR, PMDD::UINT_T, PMDA::READ_ONLY, {BLOSC_BLOSCLZ, BLOSC_LZ4, BLOSC_LZ4HC, BLOSC_SNAPPY, BLOSC_ZLIB, BLOSC_ZSTD});
+  add_status_param_metadata(CONFIG_BLOSC_THREADS, PMDD::UINT_T, PMDA::READ_ONLY, 1, ParamMetadata::MAX_UNSET);
+  add_status_param_metadata(CONFIG_BLOSC_LEVEL, PMDD::INT_T, PMDA::READ_ONLY, 0, 9);
+  add_status_param_metadata(CONFIG_BLOSC_SHUFFLE, PMDD::UINT_T, PMDA::READ_ONLY, {BLOSC_NOSHUFFLE, BLOSC_SHUFFLE, BLOSC_BITSHUFFLE});
```
Ah. These shouldn't be here and they shouldn't be returned in `status()`. Could you remove those instead of adding these?

ajgdls reviewed (changes requested):
In the cases of enums (or a set of uint options) can you set the min to 0 and the max to the max value, rather than min and max of an Integer?

Ohisemega committed:
Remove Status Parameter Metadata

Remove Status Parameter Metadata.
Remove Status Parameter Metadata Test from unit test of BloscPluginTest.
Expose derived implementations of the FrameProcessorPlugin public virtual methods.

Ohisemega committed:
Replace BloscPlugin 'Allowed Values' Entry

Replace BloscPlugin 'allowed values' entry with min and max parameters.

ajgdls commented on cpp/frameProcessor/src/BloscPlugin.cpp:
```diff
@@ -55,6 +55,16 @@ void create_cd_values(const BloscCompressionSettings& settings, std::vector<unsi
 BloscPlugin::BloscPlugin() :
 current_acquisition_(""), data_buffer_ptr_(NULL), data_buffer_size_(0)
 {
+  add_config_param_metadata(CONFIG_BLOSC_COMPRESSOR, PMDD::UINT_T, PMDA::READ_WRITE, {BLOSC_BLOSCLZ, BLOSC_LZ4, BLOSC_LZ4HC, BLOSC_SNAPPY, BLOSC_ZLIB, BLOSC_ZSTD});
```
On second thoughts I would expect that allowed_values would override any min and max.  We should probably remove them for the case of allowed_values if we can do that.  I think that should be a separate PR once we get these changes merged in.

ajgdls commented on cpp/frameProcessor/src/BloscPlugin.cpp:
```diff
@@ -55,6 +55,11 @@ void create_cd_values(const BloscCompressionSettings& settings, std::vector<unsi
+  add_config_param_metadata(CONFIG_BLOSC_COMPRESSOR, PMDD::UINT_T, PMDA::READ_WRITE, 0, 5);
+  add_config_param_metadata(CONFIG_BLOSC_THREADS, PMDD::UINT_T, PMDA::READ_WRITE, 1, ParamMetadata::MAX_UNSET);
+  add_config_param_metadata(CONFIG_BLOSC_LEVEL, PMDD::INT_T, PMDA::READ_WRITE, 0, 9);
+  add_config_param_metadata(CONFIG_BLOSC_SHUFFLE, PMDD::UINT_T, PMDA::READ_WRITE, 0, 2);
```
Yes apologies @Ohisemega for not explaining enough.  We should have allowed_values for these parameters but we should also have correct min and max for those allowed values if we cannot simply remove min and max right now.
So if allowed values is [0,1,2,3,4,5] then there should either be no min and max, or if that is not currently possible min should be 0 and max should be 5.

Ohisemega committed:
Fix Bug!

Change check from '==' to '!=' to detect unset MIN/MAX values in
add_metadata() method.

Ohisemega committed:
Revert BloscPlugin shuffle and compressor Metadata

Revert BloscPlugin 'shuffle' and 'compressor' Metadata parameters to allowed_values from min/max range.

Ohisemega committed:
Implement  FileWriterPlugin Metadata Fix Issue 408 (#424)

* Add metadata to file writer plugin

* Update FileWriterPlugin add_*_metadata() API Calls

* Remove Duplicate Aliases From FileWriterPlugin

* Update FileWriterPlugin

Remove allowed_values vector for FileWriterPlugin configuration metadata with BOOL_T datatypes.
Add default constructor to ParamMetadata class.
Replace std::unordered_map::emplace() calls in add_*_param_meadata() function with the assignment operator to allow values overriding.
Change STATUS_LAST_CHANCE static string in FileWriterPlugin class to STATUS_LAST_CREATE.

* Remove Duplicate Metadata Parameter

* Implement Configuration Metadata for ParameterAdjustmentPlugin, SumPlugin, and OffsetAdjustmentPlugin Fix Issue 408 (#425)

* Add metadata to offset and param adjust plugins

* Remove Redundant Aliases

Update add_*_param_metadata() calls to use correct namespaces.

* Implement Metadata GapFillPlugin, KafkaPlugin, and LiveViewPlugin Fix 408 (#429)

* Implement metadata for gap fill, kafka, live view plugins

* Update GapFillPlugin and LiveViewPlugin

Remove redundant aliases.
Update add_config_param_metadata() calls to utilize new API.

* Update KafkaProducerPlugin

Update KafkaProducerPlugin library add_*_param_metadata() calls.

* Updated config metadata to bool

Ohisemega committed:
Fix Bug

Change max_ argument default parameter to MAX_UNSET from MIN_UNSET in FrameProcessorPlugin::add_*_param_meadata() function overloads!
